src/odemis/gui/comp/overlay/shapes.py

```diff
@@ -92,7 +99,41 @@ def move_to(self, pos: Union[Tuple[float, float], Vec]):
 
     @abstractmethod
     def refresh(self):
-        """Refresh the shape's points VA after copy or move_to operation."""
+        """Refresh the shape's points VA after copy, move_to or restore_state operation."""
+        pass
+
+    def undo(self):
+        """Undo the last action."""
+        if self.undo_stack:
+            state = self.undo_stack.pop()
+            self.redo_stack.append(state)
+            self.restore_state(state)
+            self.cnvs.request_drawing_update()
+        else:
+            logging.warning(
+                "No undo action for %s at %s.", self.__class__.__name__, hex(id(self))
+            )
+
+    def redo(self):
+        """Redo the last undone action."""
```

Comment:
Shouldn't the redo stack be automatically emptied when `save_state` is called? Admittedly, I think that it works fine as it is, but my impression is that most software have this behaviour, so it would avoid "surprises" for the user.

Reply:
Good catch. I had this implemented and for some reason its not here in the PR.